frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ MaybeSerializeDeserialize
+			+ Zero;
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+		type Convert: Convert<Self::BlockNumber, Self::Balance>;
+		type RelayChainAccountId: Parameter + MaybeSerializeDeserialize + Into<AccountId32> + Ord;
+		#[pallet::constant]
+		type InitialPayment: Get<Perbill>;
+		#[pallet::constant]
+		type VestingPartition: Get<Self::BlockNumber>;
+		#[pallet::constant]
+		type Prefix: Get<&'static [u8]>;
+	}
+
+	#[pallet::storage]
+	pub type Rewards<T: Config> =
+		StorageMap<_, Blake2_128Concat, RemoteAccountOf<T>, RewardOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_rewards)]
+	pub type TotalRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn claimed_rewards)]
+	pub type ClaimedRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_contributors)]
+	pub type TotalContributors<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn vesting_block_start)]
+	pub type VestingBlockStart<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		pub fn initialize(origin: OriginFor<T>) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			ensure!(!VestingBlockStart::<T>::exists(), Error::<T>::AlreadyInitialized);
+			let current_block = frame_system::Pallet::<T>::block_number();
+			VestingBlockStart::<T>::set(current_block);
+			Ok(())
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn claim_from_ethereum(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+			ethereum_address: EthereumAddress,
+			proof: EcdsaSignature,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let data = reward_account.using_encoded(to_ascii_hex);
+			let signer = ethereum_recover(T::Prefix::get(), &proof, &data)
+				.ok_or(Error::<T>::InvalidProof)?;
+			// NOTE(hussein-aitlahcen): ask for the address even if we can extract it
+			// from the signature as we want to differentiate invalid proof from non-existant
+			// contributors
+			ensure!(signer == ethereum_address, Error::<T>::InvalidProof);
+			let remote_account: RemoteAccountOf<T> = RemoteAccount::Ethereum(ethereum_address);
+			let claimed = Self::do_claim(remote_account.clone(), &reward_account)?;
+			Self::deposit_event(Event::Claimed { remote_account, reward_account, amount: claimed });
+			Ok(())
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn claim_from_relay(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+			relay_account: T::RelayChainAccountId,
+			proof: MultiSignature,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
```

Comment:
not sure why this to be signed, as proof is signature

Reply:
we definitely don't want this to be signed